examples/cluster/plot_kmeans_silhouette_analysis.py

```diff
@@ -0,0 +1,140 @@
+"""
+===============================================================================
+Selecting the number of clusters with silhouette analysis on KMeans clustering
+===============================================================================
+
+Silhouette analysis can be used to study the separation distance between the
+resulting clusters. The silhouette plot displays a measure of how close each
+point in one cluster is to points in the neighboring clusters and thus provides
+a way to assess parameters like number of clusters visually. This measure has a
+range of [-1, 1].
+
+Silhoette coefficients (as these values are referred to as) near +1 indicate
+that the sample is far away from the neighboring clusters. A value of 0
+indicates that the sample is on or very close to the decision boundary between
+two neighboring clusters and negative values indicate that those samples might
+have been assigned to the wrong cluster.
+
+In this example the silhouette analysis is used to choose an optimal value for
+``n_clusters``. The silhouette plot shows that the ``n_clusters`` value of 3, 5
+and 6 are a bad pick for the given data due to the presence of clusters with
+below average silhouette scores and also due to wide fluctuations in the size
+of the silhouette plots. Silhouette analysis is more ambivalent in deciding
+between 2 and 4.
+
+Also from the thickness of the silhouette plot the cluster size can be
+visualized. The silhouette plot for cluster 0 when ``n_clusters`` is equal to
+2, is bigger in size owing to the grouping of the 3 sub clusters into one big
+cluster. However when the ``n_clusters`` is equal to 4, all the plots are more
+or less of similar thickness and hence are of similar sizes as can be also
+verified from the labelled scatter plot on the right.
+"""
```

Comment:
I sometimes wonder whether it's worth reminding the reader that while these things are easy to verify in the 2d case, it's much harder to visualise in high dimensions, for which the silhouette provides more advantage. But I'm not really certain that needs to be said, or how to say it, so thanks for what you've done.